File: 2025/day_03/go/main.go
package main

import (
	"fmt"
	"os"
	"strconv"
	"strings"
)

func read_file_to_string(inputfile string) []string {
	contents, err := os.ReadFile(inputfile)
	contents_string := strings.Split(string(contents), "\n")
	if err != nil {
		panic(fmt.Sprintf("Couldn't read inputfile: %s", inputfile))
	}
	return contents_string
}

func find_battery(sequence string, start_idx int, end_index int) (int, int) {
	max_joltage := 0
	max_joltage_idx := 0
	fmt.Printf("%s, %d, %d\n", sequence[start_idx:end_index], start_idx, end_index)
	for idx, battery := range sequence[start_idx:end_index] {
		if int(battery-'0') > max_joltage {
			max_joltage = int(battery - '0')
			max_joltage_idx = idx
		}
	}
	return max_joltage, max_joltage_idx + start_idx
}

func main() {
	contents := read_file_to_string("../testinput.input")

	// part 1
	// number_batteries := 2
	// part 2
	number_batteries := 12
	total_output := 0
	for _, line := range contents {
		var battery_array []int
		combined_joltage := ""
		battery_value := 0
		battery_idx := -1

		length := len(line)
		for idx := range number_batteries {
			battery_value, battery_idx = find_battery(line, battery_idx+1, (length - (number_batteries - 1 - idx)))
			battery_array = append(battery_array, battery_value)
		}
		for _, battery := range battery_array {
			combined_joltage += strconv.Itoa(battery)
		}
		combined_joltage_int, err := strconv.Atoi(combined_joltage)
		if err != nil {
			panic("Couldn't convert string to int")
		}
		total_output += combined_joltage_int
		fmt.Printf("%d, %d, %s\n", combined_joltage_int, total_output, line)
	}
	fmt.Println(total_output)

}

File: 2025/day_03/rust/src/main.rs
use std::fs::read_to_string;

fn read_file(filepath: String) -> Vec<String> {
    let contents: Vec<String> = read_to_string(filepath.clone())
        .expect(&format!(
            "The file: {} couldn't be converted to string",
            filepath
        ))
        .split("\n")
        .filter(|s| !s.trim().is_empty())
        .map(|s| s.to_string())
        .collect();
    return contents;
}

fn find_battery(sequence: String, start_idx: usize, end_index: usize) -> (i64, i32) {
    let battery_list = sequence.chars();
    let mut joltage: i64 = 0;
    let mut max_joltage: i64 = 0;
    let mut max_joltage_idx: usize = 0;
    println!("{}, {}", start_idx, end_index);
    for idx in start_idx..end_index {
        joltage = i64::from(battery_list.clone().nth(idx).unwrap().to_digit(10).unwrap());
        if joltage > max_joltage {
            max_joltage = joltage;
            max_joltage_idx = idx;
        }
    }
    return (max_joltage, (max_joltage_idx + start_idx) as i32);
}

fn main() {
    let contents = read_file("../testinput.input".to_string());

    // part 1
    // let number_batteries = 2
    // part 2
    let number_batteries = 12;
    let mut total_output: i64 = 0;

    for line in contents {
        let mut battery_array: Vec<i64> = vec![];
        let mut combined_joltage = "".to_string();
        let mut battery_value = 0;
        let mut battery_idx: i32 = -1;

        let length = line.len();
        for idx in 0..number_batteries - 1 {
            (battery_value, battery_idx) = find_battery(
                line.clone(),
                (battery_idx + 1) as usize,
                length - (number_batteries - 1 - idx),
            );
            battery_array.push(battery_value);
        }

        for battery in battery_array {
            combined_joltage = format!("{}{}", combined_joltage, battery.to_string());
        }
        // TODO from here
        total_output += combined_joltage.parse::<i64>().unwrap();
        println!("{}, {}, {}", combined_joltage, total_output, line)
    }
    println!("{}", total_output)
}
